src/v/cloud_storage/cache_service.cc

```diff
@@ -57,6 +57,38 @@ cache::cache(
   , _cnt(0)
   , _total_cleaned(0) {}
 
+ss::future<>
+cache::recursive_delete_empty_directory(const std::string_view& key) {
+    gate_guard guard{_gate};
+    std::filesystem::path path(key);
+
+    while (path != _cache_dir) {
+        try {
+            vassert(
+              std::string_view(path.native()).starts_with(_cache_dir.native()),
```

Comment:
this makes me nervous because it appears to be quite sensitive to format of the path, rather than the semantic meaning of the path (e.g. string operations rather than path operations). some options might be:

1. ensure that path is an absolute path in canonical form (e.g. all `..` etc... removed) and I think std::filesystem has primitives for that
2. another option would be to walk the directory hierarchical downward instead of upward looking for empty directory as a garbage collection process which would allow the GC process to start at a known point in the file system hierarchy.